numpyro/util.py

```diff
@@ -309,3 +311,190 @@ def soft_vmap(fn, xs, batch_ndims=1, chunk_size=None):
+        }
+
+    # infer plate structure
+    # TODO: subset relation might not always hold, raise exception if not
+    plate_data = {}
+    for plate1, plate2 in list(itertools.permutations(plate_groups, 2)):
+        if plate1 is None or plate2 is None:
+            continue
+
+        if set(plate_groups[plate1]) < set(plate_groups[plate2]):
+            plate_data[plate1] = {'parent': plate2}
+
+    for plate in plate_groups:
+        if plate is None:
+            continue
+
+        if plate not in plate_data:
+            plate_data[plate] = {'parent': None}
+
+    # infer RV edges
+    edge_list = []
+    for target, source_list in model_relations['sample_sample'].items():
+        edge_list.extend([(source, target) for source in source_list])
+
+    return {
+        'plate_groups': plate_groups,
+        'plate_data': plate_data,
+        'node_data': node_data,
+        'edge_list': edge_list,
+    }
+
+
+def render_graph(graph_specification):
+    """
+    Create a graphviz object given a graph specification.
+    """
+    # TODO: plate_graph_dict and plate_data assume that "deepest" plates come first. This will break!
+
+    try:
+        import graphviz  # noqa: F401
+    except ImportError as e:
+        raise ImportError(
+            'Looks like you want to use graphviz to render your model. '
+            'You need to install `graphviz` to be able to use this feature. '
+            'It can be installed with `pip install graphviz`.'
+        ) from e
+
+    plate_groups = graph_specification['plate_groups']
+    plate_data = graph_specification['plate_data']
+    node_data = graph_specification['node_data']
+    edge_list = graph_specification['edge_list']
+
+    graph = graphviz.Digraph()
+
+    # add plates
+    # TODO: order may not always be as expected (parents before children)
+    plate_graph_dict = {
+        plate: graphviz.Digraph(name=f'cluster_{plate}')
+        for plate in plate_groups
+        if plate is not None
+    }
+    for plate, plate_graph in plate_graph_dict.items():
+        plate_graph.attr(label=plate, labeljust='r', labelloc='b')
+
+    # add nodes
+    for plate, rv_list in plate_groups.items():
+        cur_graph = graph if plate is None else plate_graph_dict[plate]
+
+        for rv in rv_list:
+            color = 'grey' if node_data[rv]['is_observed'] else 'white'
+            cur_graph.node(
+                rv, label=rv, shape='circle', style='filled', fillcolor=color
+            )
+
+    for plate, data in plate_data.items():
+        parent_plate = data['parent']
+
+        if parent_plate is None:
+            graph.subgraph(plate_graph_dict[plate])
+        else:
+            plate_graph_dict[parent_plate].subgraph(plate_graph_dict[plate])
+
+    # add edges
+    for source, target in edge_list:
+        graph.edge(source, target)
+
+    # return whole graph
+    return graph
+
+
+def render_model(model, *args, filename=None, **kwargs):
```

Comment:
I think it is better to use the pattern `render_model(model, model_args=(), model_kwargs={}, filename=None, num_tries=10)`, so we can freely add more options in the future.

Reply:
Good point!

I set the default values for `model_*` to `None` (see https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments).

Reply:
agreed!